### ISSUE TITLE
Failed to reserve range: error running Wine tests

### ISSUE THREAD

**Issue by hughbe** [Microsoft/BashOnWindows#2258]
* Your Windows build number:  (Type `ver` at a Windows Command Prompt)

Microsoft Windows [Version 10.0.14393]

* What you're doing and what's happening:  (Copy&paste specific commands and their output, or include screen shots)

First, build Wine - awesome that it compiles!!
```
git clone https://github.com/wine-mirror/wine.git
cd wine
./configure
make depend
make
```

Secondly, run the tests
```
cd dlls/gdiplus/test
make font.ok
```

* What's wrong / what should be happening instead:
I'd expect the tests to run, but:

```
hugh@DESKTOP-HJANDGF:/mnt/c/Users/hugh/Documents/WSL/wine/dlls/gdiplus/tests$ make font.ok
../../../tools/runtest -q -P wine -T ../../.. -M gdiplus.dll -p gdiplus_test.exe.so font && touch font.ok
preloader: Warning: failed to reserve range 00007ffffe000000-00007fffffff0000
wine client error:0: recvmsg: Invalid argument
make: *** [font.ok] Error 1
hugh@DESKTOP-HJANDGF:/mnt/c/Users/hugh/Documents/WSL/wine/dlls/gdiplus/tests$
```

**Comment by sunjoong** [comment]
@hughbe - Here is my output of `make font.ok`;
```
root@SUNJOONG-DESKTOP:~/wine/dlls/gdiplus/tests# make font.ok
../../../tools/runtest -q -P wine -T ../../.. -M gdiplus.dll -p gdiplus_test.exe.so font && touch font.ok
preloader: Warning: failed to reserve range 00007ffffe000000-00007fffffff0000
preloader: Warning: failed to reserve range 00007ffffe000000-00007fffffff0000
preloader: Warning: failed to reserve range 00007ffffe000000-00007fffffff0000
preloader: Warning: failed to reserve range 00007ffffe000000-00007fffffff0000
preloader: Warning: failed to reserve range 00007ffffe000000-00007fffffff0000
preloader: Warning: failed to reserve range 00007ffffe000000-00007fffffff0000
preloader: Warning: failed to reserve range 00007ffffe000000-00007fffffff0000
preloader: Warning: failed to reserve range 00007ffffe000000-00007fffffff0000
fixme:gdiplus:GdipGetLineSpacing ignoring style
root@SUNJOONG-DESKTOP:~/wine/dlls/gdiplus/tests# make font.ok
make: 'font.ok' is up to date.
root@SUNJOONG-DESKTOP:~/wine/dlls/gdiplus/tests#
```

- I'm using on Microsoft Windows [Version 10.0.15063]
- I copied timesbd.ttf,  timesbi.ttf,  timesi.ttf and times.ttf to /usr/share/fonts/truetype/msttcorefonts directory.
- I was using `--enable-win64` option when configure.

**Comment by therealkenc** [comment]
Yes Wine still isn't 64-bit clean, so at the end of the day you are going to hard block on that. #228, User Voice is [here](https://wpdev.uservoice.com/forums/266908-command-prompt-console-bash-on-ubuntu-on-windo/suggestions/13377507-please-add-32-bit-elf-support-to-the-kernel). Bonus points for the "Wine Doesn't Run" PSA 🙄.

As a separate matter it might be worth an `strace` on the `recvmsg()` `EINVAL` on it's own merits. Although, given the mass hemorrhaging the code is going to experience anyway, that is possibly/probably a secondary result of other fails - the invalid argument is probably the handle itself.

**Comment by sunjoong** [comment]
@therealkenc - Thanks. FIY, advapi32 is for "Advanced Services API for 32bit systems." As noticing that, I thought more investigation might not be needed :)

**Comment by CherryDT** [comment]
@sunjoong However the name for advapi32.dll in Windows came from the days when 16-bit was standard and 32-bit was the "new thing". The name was kept when the transition to 64-bit was made, e.g. your `C:\Windows\System32\advapi32.dll` file will actually be 64-bit on your 64-bit Windows machine.

**Comment by woachk** [comment]
Try to run without the wine-preloader.

**Comment by hughbe** [comment]
@woachk how do I do that - thanks

**Comment by woachk** [comment]
@hughbe you run Wine without the preloader by simply deleting wine-preloader.

**Comment by hughbe** [comment]
Beaut - I built wine from source, so I'll let ya know how it goes

**Comment by woachk** [comment]
@therealkenc 64-bit Wine is 64-bit clean.

**Comment by sunjoong** [comment]
@CherryDT - Oh... I didn't notice that; Thank you.
@hughbe - But, I'm still not expecting wine works well on WSL. Quote from https://wiki.archlinux.org/index.php/wine ;"If you are running a 64-bit system, you will need to enable the Multilib repository first."

**Comment by woachk** [comment]
@sunjoong that's because most Windows apps are 32-bit. Wine 64-bit by itself only runs 64-bit Windows applications.

**Comment by therealkenc** [comment]
> @ therealkenc 64-bit Wine is 64-bit clean.

Correct; I should have been more pedantic. If someone wants to build a 64-bit clean version of Wine (ie, including Wine's dependencies and without multilib), post a link to a gist of a multithreaded strace of a Wine run on a Windows x86-64 PE hello world app. Maybe it will work. Maybe it will block on something else. Of course, you won't be able to run any Windows binaries that end up using Wow64 under the hood.

**Comment by woachk** [comment]
@sunjoong that's the Wine services host.

**Comment by woachk** [comment]
Wine works fine, I tried it some months ago afaik.

**Comment by sunjoong** [comment]
@woachk - I mean that's in C:\windows\system32 directory, not in wine source directory there's a new compiled services.exe file.

**Comment by woachk** [comment]
@sunjoong no, that references the services.exe in .wine/drive_c/windows/system32

**Comment by therealkenc** [comment]
Interesting. Wine64 does run on WSL. _Who knew?_ That's Ralph in an VcXsrv window. I get the "failed to reserve range" warning, but it's a soft fail. There's some other quirks, but it runs. I'll follow on with an `strace` on the fails once I take a look.

![wsl-ralph](https://user-images.githubusercontent.com/7049700/27602161-af1818b0-5b26-11e7-919c-c09bfe459a47.png)

**Comment by woachk** [comment]
@therealkenc I said it a long time ago :)

**Comment by therealkenc** [comment]
So here is the missing surface, but as woachk points out none of them are blockers:
```
[...]
prctl(PR_SET_PTRACER, 30157) = -1 EINVAL (Invalid argument)
[...]
mmap(0x14002d000, 399360, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED, 10, 0x400) = -1 
    EINVAL (Invalid argument)
```
That mmap is the source of the warning.
